MANUAL.md

```diff
@@ -1600,158 +1692,748 @@ if (field.fieldKind == "map") {
+
+  // Set the default host for this service with our custom option.
+  option (customoptions.default_host) = "https://demo.connectrpc.com/";
+
+  // ...
+}
+```
+
+Custom options are extensions to one of the options messages defined in
+`google/protobuf/descriptor.proto`. You can learn more about them in the [language guide][custom-options].
+The above option can be defined like this:
+
+```protobuf
+// customoptions/default_host.proto
+syntax = "proto3";
+import "google/protobuf/descriptor.proto";
+package customoptions;
+
+extend google.protobuf.ServiceOptions {
+  // We extend the ServiceOptions message, so that other proto files can import
+  // this file, and set the option on a service declaration.
+  optional string default_host = 1001;
+}
+```
+
+First, you need to generate code for the custom option. This will result in a file named
+`customoptions/default_host_pb.ts` with a new export: our [extension](d#extensions):
+
+```ts
+import { ServiceOptions, Extension } from "@bufbuild/protobuf";
+
+export const default_host: Extension<ServiceOptions, string> = ...
+```
+
+Now we can utilize this extension to read custom options in our plugin:
+
+```ts
+import type { GeneratedFile } from "@bufbuild/protoplugin";
+import { type DescService, getOption, hasOption } from "@bufbuild/protobuf";
+import { default_host } from "./gen/customoptions/default_host_pb.js";
+
+function generateService(service: DescService, f: GeneratedFile) {
+  // Let's see if our option was set:
+  if (hasOption(service, default_host)) {
+    const value = getOption(service, default_host); // "https://demo.connectrpc.com/"
+    // Our option was set, we can use it here.
+  }
+}
+```
+
+Custom options can be set on any Protobuf element. They can be simple singular
+string fields as the one above, but also repeated fields, message fields, etc.
+
+Take a look at our [plugin example](./packages/protoplugin-example/)
+to see the custom option above in action and run the code yourself.
+
+## Testing
+
+We recommend testing generated code just like handwritten code. Identify a representative Protobuf file for your use
+case, generate code, and then run tests against the generated code. If you implement your own generator functions for
+the `js` and `dts` targets, we also recommend running all tests against both.
+
+## Examples
+
+For a small example of generating a Twirp client based on a simple service definition, take a look at [protoplugin-example](https://github.com/bufbuild/protobuf-es/tree/main/packages/protoplugin-example).
+
+Additionally, check out [protoc-gen-es](https://github.com/bufbuild/protobuf-es/tree/main/packages/protoc-gen-es), which is the official code generator for Protobuf-ES.
+
+## Examples
+
+> **TODO** packages/protobuf-example contains a runnable example that uses the same example.User definition we've been using in this doc. We should link to it.
+
+## Migrating to Protobuf-ES
+
+TODO: Need to update migration.md and move to this doc
```

Comment:
migrating.md needs extensive updates and should also have a section about migrating from Protobuf-ES v1 to v2.

Can we remove this section and TODO for now and do that separately?